src/samplers/cpu/mod.rs

```diff
@@ -1,173 +1,288 @@
+use crate::samplers::Common;
+use crate::Sampler;
 
-use failure::Error;
-use logger::*;
-use metrics::*;
-use time;
+mod config;
+mod stat;
 
-use std::collections::HashMap;
-use std::fs::File;
-use std::io::{BufRead, BufReader};
-use std::sync::Arc;
+pub use config::*;
+pub use stat::*;
 
-const PROC_STAT: &str = "/proc/stat";
-
-// reported percentiles
-pub const PERCENTILES: &[Percentile] = &[
-    Percentile::p01,
-    Percentile::p1,
-    Percentile::p10,
-    Percentile::p25,
-    Percentile::p50,
-    Percentile::p75,
-    Percentile::p90,
-    Percentile::p99,
-    Percentile::Maximum,
-];
+#[cfg(not(feature = "perf"))]
+struct PerfCounter {}
 
+#[allow(dead_code)]
 pub struct Cpu {
     common: Common,
-    nanos_per_tick: u64,
-}
-
-struct ProcStat {
-    cpu_total: HashMap<Statistic, u64>,
+    perf_counters: CHashMap<CpuStatistic, Vec<PerfCounter>>,
+    tick_duration: u64,
 }
 
-fn read_proc_stat() -> ProcStat {
-    let file = File::open(PROC_STAT)
-        .map_err(|e| debug!("could not read {}: {}", PROC_STAT, e))
-        .unwrap();
-    let mut file = BufReader::new(file);
-    parse_proc_stat(&mut file)
+pub fn nanos_per_tick() -> u64 {
+    let ticks_per_second = sysconf::raw::sysconf(sysconf::raw::SysconfVariable::ScClkTck)
+        .expect("Failed to get Clock Ticks per Second") as u64;
+    SECOND / ticks_per_second
 }
 
-fn parse_proc_stat<T: BufRead>(reader: &mut T) -> ProcStat {
-    let mut ret = ProcStat {
-        cpu_total: HashMap::new(),
-    };
-    for line in reader.lines() {
-        let line = line.unwrap();
-        let parts: Vec<&str> = line.split_whitespace().collect();
-        if parts[0] == "cpu" && parts.len() == 11 {
-            ret.cpu_total
-                .insert(Statistic::User, parts[1].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Nice, parts[2].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::System, parts[3].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Idle, parts[4].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Irq, parts[6].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Softirq, parts[7].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Steal, parts[8].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Guest, parts[9].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::GuestNice, parts[10].parse().unwrap_or(0));
+impl Cpu {
+    pub fn spawn(config: Arc<Config>, metrics: Arc<Metrics<AtomicU32>>, handle: &Handle) {
+        if let Ok(mut cpu) = Cpu::new(config.clone(), metrics) {
+            handle.spawn(async move {
+                loop {
+                    let _ = cpu.sample().await;
+                }
+            });
+        } else if !config.fault_tolerant() {
```

Comment:
Doesn't have to be part of this PR, but it seems like this could be made shorter if it's a common pattern.

Reply:
This I would defer to another PR. There might be a few patterns we can reduce and improve on